### PR TITLE
Add priority-based blending to reflection probes.

### LINKED ISSUE #8167: Reflection Probes priority over other probes
### Describe the project you are working on

3D lighting

### Describe the problem or limitation you are having in your project

Its really common to have a general reflection probe for a part of a level and small reflection probes for small details or highly reflection props in order to have acuarte reflections

The problems is that Godot always blends the reflection probes together:

image

This can be mitigated once this PR is merged https://github.com/godotengine/godot/pull/50196 but if only mask out the reflection from a mesh, but sometimes (specially for objects that are moving) you want a probe to have a priority over other instead of a blend.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

It would be really useful to have a priority integer property in the reflection probe so when probes fully overlap it wont blend them together, instead reflecting based on the priority of the reflection probe. 
With this, maybe a blending distance would be helpful for a transition but out of that distance, the priority will take the effect.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

-

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Nope

### Is there a reason why this should be core and not an add-on in the asset library?

Reflection Probes are core

### LINKED ISSUE #101074: Reflection erroneously shows sky when blending between probes
### Tested versions

- Reproducible (and can be discovered) in v4.4.dev7.official [46c8f8c5c], possibly exposed by #99958

### System information

Godot v4.4.dev7 - Fedora Linux 41 (Workstation Edition) on Wayland - X11 display driver, Multi-window, 1 monitor - Vulkan (Forward+) - dedicated AMD Radeon RX 6700 XT (RADV NAVI22) - AMD Ryzen 7 5700X 8-Core Processor (16 threads)

### Issue description

I discovered this issue when playing around with dev 7 to see how well #99958 works and discovered this bug where the reflection from probes show the sky when blending between probes.

https://github.com/user-attachments/assets/33feccc9-542b-4946-8669-2b8c3330ebc9

### Steps to reproduce

1. Set up a scene with overlapping probes with the sky occluded
2. Add in something with 0 roughness and 1 metallic

### Minimal reproduction project (MRP)

[probe-blending.zip](https://github.com/user-attachments/files/18300639/probe-blending.zip)

### PR DESCRIPTION
**Pull request by lander-vr**
Supersedes #89408
Closes https://github.com/godotengine/godot-proposals/issues/8167
Fixes: https://github.com/godotengine/godot/issues/101074

This PR implements size-based priority blending by sorting reflection indexes based on the extents size before their contribution is computed in reflection_process. This sorting step is necessary to allow for correct blending between probes, where the correct priorities are respected in the blending areas. It also allows us to do an early-out for consecutive probes if the accum alpha channels > 1, skipping reflection/ambient calculations. Sorting is supported for up to 32 probes overlapping on a pixel.

The behavior is following the UX of Unreal Engines reflection captures because:
- A manual priority int introduces a significant extra workload on artists, especially when using a nested-probe workflow which is standard practice.
- Fully blending reflection probes is not favorable, since it introduces duplicate reflections.
- This UX is fast and effective.

The reasoning for blending based on size, as opposed to exposing a manual priority int, is outlined in more detail here: https://github.com/godotengine/godot/pull/89408#issuecomment-2505587939

This PR is only relevant for forward+ and mobile, since compatibility has strong limitations on reflection probes. I don't think it makes sense to implement these changes there.

It makes sense to merge #99958 before this pr, since the current blending logic on master doesn't allow probes to have 100% opacity, so it doesn't work great with these changes since you always still blend with lower priority probes. It also makes sense to have control over blending distances for this PR.

For the screenshots of this pr I increased the blending strength temporarily since it makes it more representative of how these changes work with #99958.
Sorting behavior:
24_12_10_12_11__godot windows editor x86_64_1KE84ngh7y

| This PR | 4.3 |
| -- | -- |
| 24_12_12_22_03__godot windows editor x86_64_CKxtThuEp9 | 24_12_10_12_50__4zDuFovebk |
| Smaller probes take priority. | All probes blend together. |

**_4.3:_**

https://github.com/user-attachments/assets/1b91a22b-dd4a-4956-8be9-864657240d59

**_This PR:_**

https://github.com/user-attachments/assets/4919dedf-ab00-4e12-89f6-c0b04d65557f

**_Interior probe behavior_**
| With other probe (This PR) | With other probe (4.3) |
| -- | -- |
| image | image |

| With sky (This PR) | With sky (4.3) |
| -- | -- |
| image | image |

### CODE REVIEW

**Comment by jcostello** [comment]
Works as intended. This make super easy to work with reflection probes and to have nice results <3

**Comment by lander-vr** [comment]
Added a check in the for loop where `reflection_process` is called to break it if both `reflection_accum.a` and `ambient_accum.a` are >= 1.

File: servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl [hunk 1]
```diff
@@ -1874,6 +1874,13 @@ void fragment_shader(in SceneData scene_data) {
 		vec3 ref_vec = normalize(reflect(-view, bent_normal));
 		ref_vec = mix(ref_vec, bent_normal, roughness * roughness);
 
+		uint sorted_reflection_indexes[32];
+		int sorted_count = 0;
+
+		for (uint i = 0; i < 32; i++) {
```
Comment (BastiaanOlij) [inline servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl:1880]: Nitpick, you're clearing the array here, but not in the mobile version of the shader. 
I think clearing it is not required as you're never reading from an index that you haven't written to first.

**Review by BastiaanOlij — APPROVED** [review]
@DarioSamo or @clayjohn can probably judge a little better whether the extra loop would have an unwanted impact and whether there is a more efficient way to do this but as in reality it's unlikely that more than 2, maybe 3, probes overlap I think it's fine.

My only question would be whether this should be an option, not always applied. Personally I've never had very good results with the current approach of just blending all overlapping probes so I'm not overly concerned that this new logic is the new way of doing it, but there might be users who wish the use the old form of blending.

Other than that. I think this is a great improvement. It could use a video showing it in action, the animations with the colored probes show well how things are working but seeing the actual end result showing a good use case, would help convince people why this is worth it.

**Comment by lander-vr** [comment]
Thanks for the review @bastiaanolij!

> in reality it's unlikely that more than 2, maybe 3, probes overlap I think it's fine.

This probably isn't true, it's likely there are more overlapping probes especially when using nested probes: It's common to add a central probe for a particular space, and add smaller probes to resolve accuracy mismatches. It is true however that it's unlikely that more than 2-3 probes will need to be computed for a given pixel. The advantage to presorting them is that we can do an early out and just skip all the irrelevant reflection_processes: The highest priority will be computed, and unless the pixel calculated is on a spot where there's blending it won't run any of the lower priority probes, so even 16 overlapping probes with this pre-sorting step should be cheaper than 16 overlapping probes on master. Only on pixels where there are multiple probes not at full strength, which realistically is only a small amount of pixels in standard situations, we need run reflection process multiple times.

> Personally I've never had very good results with the current approach of just blending all overlapping probes

I think the old approach makes the probes practically unusuable, which is an experienced I've heard echoed by colleagues using Godot. Personally I don't think the old approach is something we should keep support for, unless there's actual vocal demand for allowing probes to fully blend for whatever reason. In the end fully blending reflections is always undesirable since it introduces pretty bad artifacts.

>  It could use a video showing it in action

_**4.3**_

https://github.com/user-attachments/assets/0f7a238f-91cd-4233-9463-9b8b99ae10f2

_**This PR**_ (Again with temporary blend adjustment, to ensure the center of the probes have full blending strength.)

https://github.com/user-attachments/assets/0d4fecb2-0645-402f-8b6c-515ef6bf4d08

This does show an issue where the sky gets blended every time there's blending between probes. You can see it here too, where in the transition zones the colors get darker:
image

This is caused by:
```glsl
if (reflections.data[ref_index].exterior) {
	reflection.rgb = mix(specular_light, reflection.rgb, reflection_blend);
}
```
To fix this this blending step needs to run after the reflection_process for-loop is finished: We should only blend in the environment after accumulating all reflection contributions.

**Comment by lander-vr** [comment]
I've moved the environment blending to happen outside of reflection_process, after the for-loop. There is no darkening in the blends anymore (Ignore the weird blending between green-yellow. This is a result of the current blending logic on master, hence you see the blue probe seeping through. This is why I adjusted that in previous screenshots 🙂)
image
And no sky leakage anymore:

https://github.com/user-attachments/assets/4919dedf-ab00-4e12-89f6-c0b04d65557f

I have removed the weight float division where we are setting the specular_light and ambient_light, since we aren't accumulating all reflection contributions indefinitely, only until accum.a is 1, so this weight-division is not needed anymore.

File: servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl (hunk 1 above)
Comment (lander-vr) [reply servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl]: Makes sense! I've removed that clearing step. :)

**Comment by jcostello** [comment]
@lander-vr Nice catch

**Comment by lander-vr** [comment]
@jcostello Interior should still work:
24_12_12_18_52__godot windows editor x86_64_0zFLIcRDRM

The issues you're seeing with blending distance aren't caused by this PR, though this PR makes them more obvious, and is exactly the reason why I have added `blend = clamp(blend * 2.0, 0.0, 1.0);` at the end of the blend calculation for all of my screenshots and showcases: The blending on master never reaches full strength, which means your probe never has full contribution over what's inside. For a probe with interior toggled on, this means that the sky will still seep through. Especially now that I've moved `reflection.rgb = mix(specular_light, reflection.rgb, reflection_blend);` to be after all reflection processes have run, if the shader sees that accum.a isn't fully saturated to 1.0, it will fill the remainder with sky reflections.

So in short: probe doesn't have full strength -> blend doesn't ever reach 1.0 -> accum.a doesn't reach 1.0 -> skylight gets blended in until accum.a reaches 1.0.

Here is me toggling the interior bool without that blend adjustment (So the PR as it has been submitted):
24_12_12_18_57__godot windows editor x86_64_GPzScixxRP

This is why it's really important to merge https://github.com/godotengine/godot/pull/99958 before this PR, they go hand in hand.

If you want to test this PR, I'd advice to make the same blend adjustment `blend = clamp(blend * 2.0, 0.0, 1.0);` on line 880 in `scene_forward_lights_inc.glsl`

**Comment by jcostello** [comment]
Thanks for the explanation. IDK if it has to do with what you mention but and issue I see in this PR compared to master is this one

Al probes have 1.0 of intensity and set to internal

This PR:
Screenshot from 2024-12-12 15-43-05

Master:
Screenshot from 2024-12-12 15-24-42

**Comment by lander-vr** [comment]
Is it possible for you to share that project, then I can have a closer look and see where exactly things are going wrong in that scene. I've tested my PR in several scenes so far and haven't come across anything like this.

**Comment by jcostello** [comment]
> Is it possible for you to share that project? Then I can have a closer look and see where exactly things are going wrong in that scene. I've tested my PR in several scenes so far and haven't come across anything like this.

I uploaded it to drive. Its 2.6GB project that I use to test things. I didnt remove any part of the project, sorry

https://drive.google.com/file/d/1EaRDUhnuV0YVxy5EPNf4RD2nXz3Oalan/view?usp=sharing

**Comment by lander-vr** [comment]
Thanks for taking the time for the upload! :)

I've opened your project with my PR and it seems completely fine on my end, regardless of whether interior is toggled or not:
image

I also checked in @Calinou's Sponza demoscene, turned down the roughness on some materials, also seems to work fine:
image

Could you try toggling the `Enable Shadow` checkbox, to force them to update? If that doesn't fix it, double check that those meshes didn't lose their lightbake with the lighting debug viewmode. Either way, I'm not convinced that that issue is caused by my PR. It seems specific to a select few objects, and I don't think I can see how any of the changes that I've made would be able to cause that.

**Comment by lander-vr** [comment]
Rebased due to https://github.com/godotengine/godot/pull/99958 getting merged.

Bistro looks fine:

https://github.com/user-attachments/assets/5ce55045-ad58-4d20-b922-7f18ac21181e

And blending looks fine here too:
24_12_13_01_03__godot windows editor x86_64_z4g5DAt8lk

**Comment by jcostello** [comment]
Seems to be working now. Maybe I did something wrong. Also I have to play with the blending distance to get this right. Im courious why with high blending distance the sky gets reflected even in `inside` probes

**Comment by lander-vr** [comment]
> Also I have to play with the blending distance to get this right.

Yes this is normal, and part of the reason why it was important to have that as an exposed setting.

> Im courious why with high blending distance the sky gets reflected even in `inside` probes.

Because you're blending in surrounding reflections/probes. Inside probes now get blended appropriately with surrounding probes instead of causing the harsh cutoff we had in the past (Or bleeding due to overlap with another probe).

https://github.com/user-attachments/assets/59c8e14f-f2ab-4bfc-8a96-bee651188fe9

| This PR | master |
| -- | -- |
| image | image |
| With other probe | With other probe | 

| This PR | master |
| -- | -- |
| image | image |
| With sky | With sky | 

@Calinou I'm wondering if it could make sense to completely get rid of the green solid box for reflection probes. Despite the opacity adjustment, it still seems super intense in specific scenes and still requires you to move your view to the inside of the probe which is really cumbersome. You can't get a nice view of how the probe is blending with the surroundings because you're either only seeing whatever is outside the probe or whatever is inside.

**Comment by lander-vr** [comment]
Rebased and resolved conflict with #100344. Tested in Bistro, Sponza, and some quick diy test-scenes with no issues.

https://github.com/user-attachments/assets/4b12524b-fd30-4b6b-86b4-57868e19ed9f

**Comment by Calinou** [comment]
> @Calinou I'm wondering if it could make sense to completely get rid of the green solid box for reflection probes. Despite the opacity adjustment, it still seems super intense in specific scenes and still requires you to move your view to the inside of the probe which is really cumbersome, it's extremely intrusive. You can't get a nice view of how the probe is blending with the surroundings because you're either only seeing whatever is outside the probe or whatever is inside.

That makes sense. We might want to do the same for VoxelGI as well, keeping the lines only.

**Comment by jcostello** [comment]
> Rebased and resolved conflict with #100344. Tested in Bistro, Sponza, and some quick diy test-scenes with no issues.
> 

Awesome, I think it works as expected

**Comment by lander-vr** [comment]
I've reworked the entire thing and it has been severely simplified and optimized.

I was worried about the cost of sorting in GLSL and found with some quick and dirty performance tests that with several overlapping probes (e.g. the Bistro scene that I set up with probes) there was a not insignificant difference between master and this PR.

I've gotten rid of the sorting step in the shader, reverted the `reflection_process()` to be executed in the main reflection loop, and removed that secondary for-loop that I added. There was already sorting happening in `light_storage.cpp` based on the depth of the probe to the camera, so I've changed it to sort based on the extents instead. This means that the probe indexes are now already sorted before they enter the shader.

From my very rudimentary performance tests, the addition of the early out (break loop if both reflection_accum.a and ambient_accum.a are >= 1.0) gives a small performance increase when there are several probes overlapping.

This is very unscientifically tested, just eyeballed with a camera in the same position basically, but I do see improvements in frametimes on my PR compared to master with several/many overlapping probes.

**Comment by lander-vr** [comment]
Preliminary implementation done for mobile, adding a sorting function in `light_storage.cpp`. It works, but sorting only happens when the probes are paired to the geometry, so changing a probes size doesn't update the sorting until `pair_reflection_probe_instances()` is called.

**Review by Calinou — COMMENTED** [review]
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/08508d2e01567446ef7cf721b39041a51e2eddfd), it works as expected.

**Testing project:** [test_pr_100241.zip](https://github.com/user-attachments/files/18154699/test_pr_100241.zip)

## Comparison

Visually, it's quite different in Forward+, but it's almost the same in Mobile. Is the visual difference intended? I'd expect Forward+ to look as close as possible to Mobile, like it did before.

### Forward+

Before | After
-|-
Screenshot_20241216_184909 png webp | Screenshot_20241216_184930 png webp

### Mobile

Before | After
-|-
Screenshot_20241216_184915 png webp | Screenshot_20241216_184939 png webp

## Benchmark

<details open>

<summary>PC specifications</summary>

- **CPU:** Intel Core i9-13900K
- **GPU:** NVIDIA GeForce RTX 4090
- **RAM:** 64 GB (2×32 GB DDR5-5800 C30)
- **SSD:** Solidigm P44 Pro 2 TB
- **OS:** Linux (Fedora 41)
</details>

All tests done in 3840×2160.

### Before

Forward+ | Mobile | Compatibility[^1]
-|-|-
1151 FPS (0.87 mspf) | 1753 FPS (0.57 mspf) | 3915 FPS (0.25 mspf)

### After

Forward+ | Mobile | Compatibility[^1]
-|-|-
1278 FPS (0.78 mspf) | 2896 FPS (0.34 mspf) | 3903 FPS (0.25 mspf)

[^1]: Only 2 probes per mesh can be rendered, so some of them were skipped during rendering.

**Comment by lander-vr** [comment]
@Calinou Thanks for the tests!

The main reason for those visual inconsistencies is due to the unorthodox setup. If you have a bunch of overlapping probes with the same size, whichever one is loaded first will be given priority since there is not a single one that takes priority by being smaller and getting sorted to the front. In reality this isn't an issue, because there is no reason to have (or for us to properly support) a probe setup as you have here. (Some more info on probe setups here https://github.com/godotengine/godot/pull/89408#issuecomment-2505587939)

If this gets merged I'll invest some time into updating the documentation around ReflectionProbes, how they work, how to use them efficiently, and how to get good result. :)

If you increase the size of that central probe you should get a much more similar result in foward+.
image

I'm a bit confused by your frame times, I see you have higher frame times using the PR compared to master, which shouldn't be the case anymore since I switched the sorting to happen on CPU, unless I guess the scene would be cpu bottlenecked. If I compare your test scene in foward+ I get:
PR: 1.37ms
Master: 1.65ms
on a 5600x and rtx2060

**Comment by lander-vr** [comment]
Moved the sorting function to be called in `fill_push_constant_instance_indices()`, which resolves the sorting not updating when changing the probes sizes, and I changed `GeometryInstanceForwardMobile` to store the RID of the probe instances instead of the forward ID, because in order to sort the probes we need access to the size property (for which we need the RID).

I'm not super convinced that this is the absolute best way, I think sorting in the pairing function makes a bit more sense still, since then we avoid sorting needlessly when the order couldn't have changed, but I couldn't figure out where to call that function so it runs when the probes are marked as dirty.

Sorting now behaves as expected on mobile. :)

**Comment by Calinou** [comment]
> I'm a bit confused by your frame times, I see you have higher frame times using the PR compared to master, which shouldn't be the case anymore since I switched the sorting to happen on CPU. If I compare your test scene in foward+ I get:

I had the Before and After tables swapped, sorry. I fixed the post accordingly.

File: servers/rendering/renderer_rd/storage_rd/light_storage.cpp [hunk 2]
```diff
@@ -1718,6 +1718,33 @@ void LightStorage::set_max_reflection_probes(const uint32_t p_max_reflection_pro
 	reflection_buffer = RD::get_singleton()->storage_buffer_create(sizeof(ReflectionData) * max_reflections);
 }
 
+void LightStorage::sort_mobile_reflection_probes(const RID *p_reflection_probe_instances, RendererRD::ForwardID *p_reflection_probes, uint32_t reflection_probe_count) {
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/storage_rd/light_storage.cpp:1721]: I think this whole function can be simplified because you have already sorted the reflection probes by size when filling the reflection probe buffer. Therefore the index stored by the ForwardID is enough to sort them by size using `forward_id_allocators[p_type].map[p_id]`. You just have to ensure that they are stored by smallest to largest index. The nice benefit of doing that is you can keep all your logic inside `fill_push_constant_instance_indices()` and you don't have to pass around references to the instances either.

Comment (lander-vr) [reply servers/rendering/renderer_rd/storage_rd/light_storage.cpp]: This makes a ton of sense, the implementation is a lot cleaner now. :)

File: servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp [hunk 3]
```diff
@@ -124,14 +124,23 @@ void RenderForwardMobile::fill_push_constant_instance_indices(SceneState::Instan
 	p_instance_data->reflection_probes[0] = 0xFFFFFFFF;
 	p_instance_data->reflection_probes[1] = 0xFFFFFFFF;
 
+	RendererRD::ForwardID sorted_reflection_probes[MAX_RDL_CULL];
+	for (uint32_t i = 0; i < p_instance->reflection_probe_count; i++) {
+		sorted_reflection_probes[i] = p_instance->reflection_probes[i];
+	}
+
+	std::sort(sorted_reflection_probes, sorted_reflection_probes + p_instance->reflection_probe_count, [&](RendererRD::ForwardID a, RendererRD::ForwardID b) {
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp:132]: We don't use std::sort, so this should use the SortArray class instead.

Comment (lander-vr) [reply servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp]: Done! I'm still unfamiliar with the SortArray class, I just followed suit with how the probe instances are sorted by size with the `ReflectionProbeInstanceSort` struct in `update_reflection_probe_buffer()`, so I hope that approach is fine here as well. :)

File: servers/rendering/renderer_rd/storage_rd/light_storage.cpp [hunk 4]
```diff
@@ -1732,11 +1732,12 @@ void LightStorage::update_reflection_probe_buffer(RenderDataRD *p_render_data, c
 		if (!rpi) {
 			continue;
 		}
-
-		Transform3D transform = rpi->transform;
+		ReflectionProbe *probe = reflection_probe_owner.get_or_null(rpi->probe);
+		Vector3 extents = probe->size / 2;
+		float probe_size = Math::sqrt(extents.x * extents.x + extents.y * extents.y + extents.z * extents.z);
```
Comment (AThousandShips) [inline servers/rendering/renderer_rd/storage_rd/light_storage.cpp:1737]:
```suggestion
		float probe_size = extents.length();
```
Or am I missing something

**Review by AThousandShips — APPROVED** [review]

File: servers/rendering/renderer_rd/storage_rd/light_storage.cpp (hunk 4 above)
Comment (lander-vr) [reply servers/rendering/renderer_rd/storage_rd/light_storage.cpp:1737]: Oh yeah, you're totally right, that makes a lot more sense 😂

**Review by clayjohn — APPROVED** [review]
Let's gooooooooooo!

This is really great work
